ntoskrnl/ke/amd64/xstate.c

```diff
@@ -0,0 +1,270 @@
+        (XState->Size != GlobalXState->Size) ||
+        (XState->AllFeatureSize != GlobalXState->AllFeatureSize))
+    {
+        /* Invalid features */
+        KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+    }
+
+    for (ULONG i = 0; i < MAXIMUM_XSTATE_FEATURES; i++)
+    {
+        if ((XState->Features[i].Size != GlobalXState->Features[i].Size) ||
+            (XState->Features[i].Offset != GlobalXState->Features[i].Offset) ||
+            (XState->AllFeatures[i] != GlobalXState->AllFeatures[i]))
+        {
+            /* Invalid features */
+            KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+        }
+    }
+}
+
+/*!
+ *   \brief Initializes the extended state configuration for the current processor
+ *
+ *   \param ProcessorNumber - Number of the current processor
+ */
+CODE_SEG("INIT")
+VOID
+NTAPI
+KiInitializeXStateConfiguration(
+    _In_ ULONG ProcessorNumber)
+{
+    if (ProcessorNumber == 0)
+    {
+        /* Processor 0: Retrieve the global configuration */
+        KiGetXStateConfiguration(&SharedUserData->XState);
+        KeXStateLength = SharedUserData->XState.AllFeatureSize;
+    }
+    else
+    {
+        /* Application processor: validate the configuration against the global one */
+        XSTATE_CONFIGURATION XState;
+        KiGetXStateConfiguration(&XState);
+        ValidateXStateConfig(&XState);
+    }
+
+    /* Enable the user mode components in XCR0 */
+    _xsetbv(0, SharedUserData->XState.EnabledFeatures);
+
+    /* Now that we have set everything up, query CPUID again to get the required
+       size based on components enabled in XCR0 */
+    CPUID_EXTENDED_STATE_MAIN_LEAF_REGS ExtStateMain;
+    __cpuidex(ExtStateMain.AsInt32,
+              CPUID_EXTENDED_STATE,
+              CPUID_EXTENDED_STATE_MAIN_LEAF);
+
+    /* Make sure our calculation adds up! We only make sure that our size is
+       large enough, as some VMs always report the full size, independent of
+       what has been enabled in XCR0 */
+    if (ExtStateMain.Ebx > SharedUserData->XState.Size)
+    {
+        KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+    }
+
+    /* Check if we have any supervisor components enabled */
+    if (SharedUserData->XState.EnabledSupervisorFeatures != 0)
+    {
+        /* Enable the supervisor components in IA32_XSS */
+        __writemsr(MSR_IA32_XSS, SharedUserData->XState.EnabledSupervisorFeatures);
+
+        /* Get the required size for features enabled in both XCR0 and IA32_XSS */
+        CPUID_EXTENDED_STATE_SUB_LEAF_REGS ExtStateSubLeaf;
+        __cpuidex(ExtStateSubLeaf.AsInt32,
+                  CPUID_EXTENDED_STATE,
+                  CPUID_EXTENDED_STATE_SUB_LEAF);
+
+        /* Check if all components fit into what we calculated */
+        if (ExtStateSubLeaf.Ebx.XSaveAreaSize > SharedUserData->XState.AllFeatureSize)
+        {
+            KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
```

Comment:
Here also you could use KeBugCheckEx to display the value of XSaveAreaSize and AllFeatureSize